File: app/src/main/java/com/example/botcoin/za/co/botcoin/utils/GeneralUtils.java
package com.example.botcoin.za.co.botcoin.utils;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.util.Base64;
import android.util.Log;
import android.widget.Toast;

import androidx.constraintlayout.widget.ConstraintLayout;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

public class GeneralUtils {
    public static String getCurrentDateTime()
    {
        String toReturn = "";
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd HH-mm-ss");
        Date now = new Date();

        Calendar calendar = Calendar.getInstance();
        calendar.setTime(now);
        now = calendar.getTime();

        toReturn = simpleDateFormat.format(now);

        return toReturn;

    }

    public static void makeToast(Context context, String message)
    {
        Toast.makeText(context, message, Toast.LENGTH_LONG).show();
    }

    public static String getAuth()
    {
        String toReturn = null;

        try
        {
            String username = ConstantUtils.KEY_ID;
            String password = ConstantUtils.SECRET_KEY;

            String auth = username + ":" + password;
            byte[] authEncBytes = Base64.encode(auth.getBytes(), 0);
            toReturn = new String(authEncBytes);
        }catch(Exception e)
        {
            Log.e(ConstantUtils.BOTCOIN_TAG, "\nError: " + e.getMessage()
                    + "\nMethod: getAuth"
                    + "\nCreatedTime: " + GeneralUtils.getCurrentDateTime());
        }
        return toReturn;
    }

    public static boolean isApiCredentialsSaved(){
        boolean toReturn = false;

        if(ConstantUtils.KEY_ID != null && ConstantUtils.SECRET_KEY != null)
        {
            toReturn = true;
        }

        return toReturn;
    }

    public static AlertDialog createAlertDialog(Context context,String title,  String message, boolean isPrompt)
    {
        AlertDialog toReturn = null;

        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        builder.setTitle(title);
        builder.setMessage(message);


        if(isPrompt)
        {
            builder.setCancelable(true);
            builder.setPositiveButton(
                "Yes",
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        dialog.cancel();
                    }
                });

            builder.setNegativeButton(
                "No",
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        dialog.cancel();
                    }
                });
        }else
        {
            builder.setCancelable(false);
            builder.setPositiveButton(
                "Okay",
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        dialog.cancel();
                    }
                });
        }

        toReturn = builder.create();

        return toReturn;

    }
}

File: app/src/main/java/za/co/botcoin/utils/GeneralUtils.kt
package za.co.botcoin.utils

import android.app.*
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Build
import android.util.Base64
import android.util.Log
import android.view.View
import com.google.android.material.snackbar.Snackbar
import com.google.zxing.BarcodeFormat
import com.google.zxing.qrcode.QRCodeWriter
import za.co.botcoin.R
import za.co.botcoin.services.FiboService
import za.co.botcoin.utils.services.sharedPreferencesService.BaseSharedPreferencesService

object GeneralUtils {
    private const val BOTCOIN_TAG = "BOTCOIN"

    fun getAuth(keyId: String, secretKey: String): String {
        try {
            val auth = "$keyId:$secretKey"
            val authEncBytes = Base64.encode(auth.toByteArray(), Base64.NO_WRAP)
            return String(authEncBytes)
        } catch (e: Exception) {
            Log.e(BOTCOIN_TAG, "Error: ${e.message} " +
                    "Method: getAuth " +
                    "CreatedTime: ${DateTimeUtils.getCurrentDateTime()}")
        }
        return ""
    }

    fun isApiKeySet(context: Context?): Boolean {
        val lunoApiKeyId = context?.let { BaseSharedPreferencesService[it, BaseSharedPreferencesService.LUNO_API_KEY_ID] }
        val lunoApiSecretKey = context?.let { BaseSharedPreferencesService[it, BaseSharedPreferencesService.LUNO_API_SECRET_KEY] }
        if (!lunoApiKeyId.isNullOrBlank() && !lunoApiSecretKey.isNullOrBlank()) {
            ConstantUtils.USER_KEY_ID = lunoApiKeyId
            ConstantUtils.USER_SECRET_KEY = lunoApiSecretKey
            return true
        }
        return true
    }

    fun notify(view: View, message: String) {
        Snackbar.make(view, message, Snackbar.LENGTH_LONG).show()
    }

    fun createAlertDialog(context: Context?, title: String, message: String, isPrompt: Boolean): AlertDialog {
        val builder = AlertDialog.Builder(context)

        builder.setTitle(title)
        builder.setMessage(message)

        if (isPrompt) {
            builder.setCancelable(true)
            builder.setPositiveButton("Yes") { dialog, _ -> dialog.cancel() }
            builder.setNegativeButton("No") { dialog, _ -> dialog.cancel() }
        } else {
            builder.setCancelable(false)
            builder.setPositiveButton("Okay") { dialog, _ -> dialog.cancel() }
        }

        return builder.create()
    }

    fun createQRCode(codeContent: String, width: Int, height: Int): Bitmap {
        val qrCodeWriter = QRCodeWriter()
        val bitMatrix = qrCodeWriter.encode(codeContent, BarcodeFormat.QR_CODE, width, height)
        val toReturn: Bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565)
        for (x in 0 until width) {
            for (y in 0 until height) {
                toReturn.setPixel(x, y, if (bitMatrix[x, y]) Color.BLACK else Color.WHITE)
            }
        }
        return toReturn
    }

    fun notify(context: Context?, title: String, message: String) {
        context?.let {
            val notificationManager = it.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val CHANNEL_ID = "BotCoin"
                notificationManager.createNotificationChannel(NotificationChannel(CHANNEL_ID, "BotCoin", NotificationManager.IMPORTANCE_DEFAULT))
                val notification = Notification.Builder(it)
                        .setContentTitle(title)
                        .setContentText(message)
                        .setSmallIcon(R.mipmap.botcoin)
                        .setChannelId(CHANNEL_ID)
                        .build()
                notificationManager.notify(0, notification)
            } else {
                val notification = Notification.Builder(it)
                        .setContentTitle(title)
                        .setContentText(message)
                        .setSmallIcon(R.drawable.botcoin)
                        .setContentIntent(PendingIntent.getActivity(it, 0, Intent(), 0))
                        .setDefaults(Notification.DEFAULT_SOUND or Notification.DEFAULT_LIGHTS or Notification.DEFAULT_VIBRATE)
                        .setAutoCancel(true)
                        .build()
                notificationManager.notify(0, notification)
            }
        }
    }

    fun runAutoTrade(context: Context) {
        if (isApiKeySet(context)) {
            val isAutoTrade = BaseSharedPreferencesService[context, BaseSharedPreferencesService.AUTO_TRADE_PREF]
            if (!isAutoTrade.isNullOrBlank() && isAutoTrade.toBoolean()) {
                if (Build.VERSION.SDK_INT >= 26) {
                    context.startForegroundService(Intent(context, FiboService::class.java))
                } else {
                    context.startService(Intent(context, FiboService::class.java))
                }
            } else {
                context.stopService(Intent(context, FiboService::class.java))
            }
        } else {
            context.stopService(Intent(context, FiboService::class.java))
        }
    }
}